File: LeetCode/Hao/137SingleNumberII.java
public class Solution {
    public int singleNumber(int[] nums) {
        int[] b = new int[32];
        
        for (int i = 0; i < nums.length; i ++) {
            int bit = 1;
            for (int j = 0; j < 32; j ++) {
                if ((bit & nums[i]) != 0) {
                    b[j] += 1;
                }
                bit <<= 1;
            }
        }
        
        int result = 0;
        int bit = 1;
        for (int i = 0; i < 32; i ++) {
            if (b[i] % 3 != 0) {
                result += bit;
            }
            
            bit <<= 1;
        }
        
        return result;
    }
}

File: LeetCode/Liang/137SingleNumberII.cpp
/**
 * Yes back to the curcuit design:
 * https://leetcode.com/discuss/54970/an-general-way-to-handle-all-this-sort-of-questions
 **/

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int one = 0, two = 0;
        for (int i = 0; i < nums.size(); ++i) {
            int tmp = (one & ~two & nums.at(i)) | (~one & two & ~nums.at(i));
            one = (~one & ~two & nums.at(i)) | (one & ~two & ~nums.at(i));
            two = tmp;
        }
        return one | two; // if it appear one time, it will be stored in one; otherwise, two;
    }
};
